### PR TITLE
Feature/halopsa node WIP

### PR DESCRIPTION
**Pull request by michael-radency**
Working on integration for HaloPSA

### CODE REVIEW

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import { INodeProperties } from 'n8n-workflow';
+
+export const clientDescription: INodeProperties[] = [
+	{
+		displayName: 'Name:',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:5]: Name is only required when creating, not when updating. So, when updating, move to a collection.

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	{
+		displayName: 'Additional Fields',
+		name: 'additionalFields',
+		type: 'collection',
+		default: {},
+		placeholder: 'Add Field',
+		displayOptions: {
+			show: {
+				operation: ['update', 'create'],
+				resource: ['client'],
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:28]: FIelds under a collection need to be ordered alphabetically

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		displayOptions: {
+			show: {
+				operation: ['create', 'update'],
+				resource: ['client'],
+			},
+		},
+	},
+	// Additional fields =============================================================
+	{
+		displayName: 'Additional Fields',
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:20]: Additional fields are called when creating but then updating it should be called Update Fields. Check the Jira node for a reference.

File: packages/nodes-base/nodes/HaloPSA/descriptions/InvoiceDescription.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		required: true,
+		displayOptions: {
+			show: {
+				operation: ['create', 'update'],
+				resource: ['invoice'],
+			},
+		},
+	},
+	// Invoice Items =============================================================
+	{
```
Comment (RicardoE105) [inline packages/nodes-base/nodes/HaloPSA/descriptions/InvoiceDescription.ts:36]: Are the invoice items required?

File: packages/nodes-base/credentials/HaloPSAApi.credentials.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			default: '',
+			description: 'An additional tenant parameter for HaloPSA hosted solution',
+		},
+		{
+			displayName: 'Scope',
+			name: 'scope',
+			type: 'string',
+			default: 'all',
+			// required: true,
+			description: 'Must include any application permissions that are required by your application',
```
[the remark this answers is not in the document]
Comment (michael-radency) [reply packages/nodes-base/credentials/HaloPSAApi.credentials.ts:81]: scope need to be set on haloPSA side, here is just a value for auth request

File: packages/nodes-base/credentials/HaloPSAApi.credentials.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+			description: 'Your Halo Authorisation Server Url',
+		},
+		{
+			displayName: 'Resource Server',
+			name: 'resourceApiUrl',
+			type: 'string',
+			default: '',
+			// required: true,
+			description:
+				'By default, the Resource server is available at *your Halo Web Applicaiton url*"/api". Each resource then has it\'s own endpoint, e.g Tickets are available at *your Halo Web Applicaiton url*"/api/tickets". Endpoints accept the HTTP GET, POST and DELETE methods depending on the resource that you are accessing.',
```
[the remark this answers is not in the document]
Comment (michael-radency) [reply packages/nodes-base/credentials/HaloPSAApi.credentials.ts:43]: it's excerpt from docs https://n8nmichael.halopsa.com/apidoc/resources, tenant used only for auth

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+			{
+				displayName: 'Include Active',
+				name: 'includeactive',
+				type: 'boolean',
+				default: true,
+				description: 'Whether to include active customers in the response',
+			},
+			{
+				displayName: 'Include Inactive',
+				name: 'includeinactive',
```
Comment (janober) [inline packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:216]: Why not in camelCase?

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+				],
+				operation: [
+					'getAll',
+				],
+			},
+		},
+		options: [
+			{
+				displayName: 'Include Active',
+				name: 'includeactive',
```
Comment (janober) [inline packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:209]: Why not in camelCase?

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts (hunk 7 above)
Comment (janober) [reply packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:216]: Is it about how API uses it?

File: packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts (hunk 8 above)
Comment (janober) [reply packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:209]: Is it about how API uses it?

Comment (michael-radency) [reply packages/nodes-base/nodes/HaloPSA/descriptions/ClientDescription.ts:209]: yes, but I'll replace it with options field and change to camelCase

**Comment by janober** [comment]
Thanks a lot. Got merged.

**Comment by janober** [comment]
Got released with `n8n@0.163.0`

**Pull request merged**